### PR TITLE
fix(search-bar): recover self-corrected filter arrays in Ask AI parser (LFE-11007)

### PR DESCRIPTION
**Pull request by nkabardin**
## TL;DR

Ask AI (the v4 search-bar natural-language filter) sometimes threw away a **correct** answer. When the model self-corrected — emitting a draft filter array, some prose, then a fixed second array — the parser's greedy `/\[[\s\S]*\]/` match spanned the *first* `[` to the *last* `]`, swallowing both arrays and the prose in between, so `JSON.parse` failed and the user just saw "couldn't build filters." This makes the parser extract **balanced top-level arrays** and prefer the model's **self-corrected (last) one**. Also extends the score-name guardrail to boolean scores. 27/27 unit tests, clean-path behavior unchanged.

## Why

Mining prod Ask AI traces (LFE-10716) turned up a real user who lost a perfect answer to exactly this: the model produced a draft, reconsidered in prose ("that column can't express starts-with…"), then emitted a corrected array — and the greedy regex made the whole thing unparseable. It's a recurring failure class for any self-correcting completion, and it's invisible to the user (looks like the feature just can't do it).

## What

- **Balanced extraction** — a new string-aware, depth-tracked scan (`extractTopLevelArrays`) collects every top-level `[…]` candidate. Brackets inside JSON string values are inert (a `]` in a value can't close an array early); escapes are handled.
- **Last-first preference** — candidates are tried whole-string first (so the existing bare-array and `{filters:[…]}` wrapper paths are byte-for-byte unchanged), then the extracted arrays **last-first**, since the self-corrected array is the intended one.
- **Boolean score-name guardrail** — the existing name-correction/validation (numeric + categorical) now also covers boolean score sets (`score_booleans` / `trace_score_booleans`), closing the same dead-filter class for boolean scores. Additive + backward compatible: an absent set means no enforcement, identical to today.

## Result

- The self-correction case now recovers the corrected filters instead of failing.
- **No masking:** the fallback path only ever carries *empty* filters (+ the raw count for telemetry), so a wrong non-empty filter set is never silently applied; dropped/unknown names still surface via `droppedCount` / `unknownScoreNames`.
- Clean-path (single array, `{filters:[…]}` wrapper) unchanged.
- **27/27** server-unit tests pass (20 pre-existing + 7 new).

<details>
<summary>Mechanism, tests, known edge, scope</summary>

**Files:** `parseFilterCompletion.ts` (+115/−26), `lib/observed-options.ts` (+4: boolean sets on `ObservedScoreNames`), `server/router.ts` (+2 optional Zod fields), `__tests__/…/parseFilterCompletion.servertest.ts` (+128).

**New tests:** self-correction regression (draft `DEBUG` + prose + corrected `ERROR` → asserts corrected wins), single-array-in-prose (unchanged), `]`-inside-a-string-value balance, brackets-but-no-valid-JSON → empty result, and three boolean cases (separator correction, unknown trace-boolean dropped + reported, pass-through when the set is absent).

**Known edge (documented, non-blocking):** an *odd/unbalanced* `"` in the inter-array prose (e.g. a stray inch-mark) opens a string state that can swallow the corrected array's brackets, in which case the guardrail-validated **draft** is applied instead of the correction. This is rare (normal model prose has balanced quotes) and is **not a regression** — the old greedy code returned nothing at all in that same case. Possible follow-up hardening: a forward-order secondary pass. Filed under the parent epic.

**Scope discipline:** OR-group / nested `{"type":"group","operator":"OR"}` tree-contract support was intentionally **not** added here — it's a separate subtask (feature, not bug). This PR stays on the flat `FilterState` contract.

Fixtures are synthesized (no real values).
</details>

Parent epic: LFE-10716 · subtask: LFE-11007.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes a bug in the Ask AI search-bar parser where a greedy regex (`/\[[\s\S]*\]/`) would span from the first `[` to the last `]` in a model completion, causing `JSON.parse` to fail whenever the model self-corrected — emitting a draft filter array, some prose, then a corrected array. The fix replaces the regex with a string-aware, depth-tracked balanced bracket scanner and tries candidates last-first so the self-corrected array wins. It also extends the existing score-name guardrail to `booleanObject` filters.

- **`extractTopLevelArrays`**: a character-by-character scanner that tracks `"..."` string state (including escape sequences), so a `]` inside a JSON string value never closes an array prematurely, and nested arrays inside filter objects are captured as part of their enclosing top-level array.
- **Last-first candidate preference**: the whole completion is still tried first (preserving the clean-path and `{filters:[…]}` wrapper paths byte-for-byte), then extracted arrays are tried last-first; the `fallback` mechanism ensures a stray zero-valid-filter array in the prose never shadows a real filter array that preceded it.
- **Boolean score guardrail**: `booleans` / `traceBooleans` added to `ObservedScoreNames`, `router.ts` Zod schema, and `SCORE_NAME_SET_BY_COLUMN`, closing the same dead-filter class for boolean scores that was already fixed for numeric and categorical.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge. The core filter extraction path is unchanged for the common cases (bare array, `{filters:[…]}` wrapper), and the new balanced-bracket scanner is thoroughly tested.

The `extractTopLevelArrays` scanner and the last-first candidate ordering are logically correct and well-covered by 27 tests. The only issue found is a telemetry inaccuracy in the `droppedCount` for a degenerate edge case where every candidate parses as JSON but no candidate yields valid filter items — the fallback `rawCount` can reflect a stray prose array's element count rather than the intended filter array's. This has no effect on the filters actually applied to queries.

`parseFilterCompletion.ts` is the only file with non-trivial logic changes and warrants the most attention, specifically the `fallback` accumulation in `parseFilterArray` and the new `extractTopLevelArrays` scanner.
</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[completion string] --> B[parseFilterArray]
    B --> C[candidates = completion + extractTopLevelArrays reversed]
    C --> D{Try each candidate}
    D --> E[JSON.parse]
    E -->|throws| D
    E -->|ok| F{raw is array?}
    F -->|no| D
    F -->|yes| G[safeParse each element]
    G --> H{kept.length > 0?}
    H -->|yes| I[return filters + rawCount]
    H -->|no| J{fallback set?}
    J -->|no| K[store as fallback rawCount=raw.length]
    K --> D
    J -->|yes| D
    D -->|exhausted| L{fallback?}
    L -->|yes| M[return fallback: filters=empty]
    L -->|no| N[return filters=empty rawCount=0]

    subgraph extractTopLevelArrays
    O[scan text char by char] --> P{in string?}
    P -->|yes| Q[handle escape + close-quote only]
    P -->|no| R{char type?}
    R -->|quote| S[set inString=true]
    R -->|open bracket at depth 0| T[record start, depth++]
    R -->|open bracket at depth > 0| U[depth++]
    R -->|close bracket| V[depth--]
    V --> W{depth==0?}
    W -->|yes| X[emit top-level array]
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[completion string] --> B[parseFilterArray]
    B --> C[candidates = completion + extractTopLevelArrays reversed]
    C --> D{Try each candidate}
    D --> E[JSON.parse]
    E -->|throws| D
    E -->|ok| F{raw is array?}
    F -->|no| D
    F -->|yes| G[safeParse each element]
    G --> H{kept.length > 0?}
    H -->|yes| I[return filters + rawCount]
    H -->|no| J{fallback set?}
    J -->|no| K[store as fallback rawCount=raw.length]
    K --> D
    J -->|yes| D
    D -->|exhausted| L{fallback?}
    L -->|yes| M[return fallback: filters=empty]
    L -->|no| N[return filters=empty rawCount=0]

    subgraph extractTopLevelArrays
    O[scan text char by char] --> P{in string?}
    P -->|yes| Q[handle escape + close-quote only]
    P -->|no| R{char type?}
    R -->|quote| S[set inString=true]
    R -->|open bracket at depth 0| T[record start, depth++]
    R -->|open bracket at depth > 0| U[depth++]
    R -->|close bracket| V[depth--]
    V --> W{depth==0?}
    W -->|yes| X[emit top-level array]
    end
```

</a>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
web/src/features/search-bar/server/parseFilterCompletion.ts:121-148
**`fallback` rawCount may reflect the wrong array in the all-fail case**

When every candidate parses as a JSON array but none yields valid filter items, the `fallback` is set from whichever candidate first hits the `kept.length === 0` branch — in iteration order that is the last text array (tried first after reversing). If that first no-valid-filter candidate happens to be a stray JSON array in the prose (e.g. `["a", "b"]`, 2 elements) while the model's intended filter array is a different candidate (e.g. 4 elements, all malformed), the returned `rawCount` is 2, so `droppedCount` is 2 rather than 4 in the caller. This is telemetry-only (filters are empty either way), but the count misrepresents how many model items were actually dropped.

In the common cases — at least one candidate has valid filters, or no candidate parses as a JSON array at all — the `fallback` is either never consulted or never set, so those paths are unaffected.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(search-bar): recover self-corrected ..."](https://github.com/langfuse/langfuse/commit/bbb58b459a83f30c27ace83e33af97eb2ba0ff04) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44809623)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: web/src/features/search-bar/server/parseFilterCompletion.ts [hunk 1]
```diff
@@ -51,44 +51,101 @@ function isEventsContractCompatible(f: FilterState[number]): boolean {
...
+    const kept: FilterState = [];
+    for (const item of raw) {
+      const result = singleFilter.safeParse(item);
+      if (result.success) kept.push(result.data);
     }
+    if (kept.length > 0) return { filters: kept, rawCount: raw.length };
+    if (fallback === null) fallback = { filters: [], rawCount: raw.length };
   }
-  return { filters: [], rawCount: 0 };
+  return fallback ?? { filters: [], rawCount: 0 };
```
[the remark this answers is not in the document]
Comment (nkabardin) [reply web/src/features/search-bar/server/parseFilterCompletion.ts:165]: Fixed Failure 1 in f8e4cd60: an explicit trailing `[]` retraction now returns empty immediately instead of falling through to the earlier draft (test added — the ERROR-draft-then-`[]` repro). Failure 2 (a later structurally-valid-but-downstream-dead array shadowing an earlier correct one) is non-regressing vs main, so I've filed it as a follow-up under LFE-10716 rather than expanding this PR.
